File: frontend/src/types/tune.ts
export interface Note {
  pitch: string;
  duration: number;
  start_time: number;
}

export interface Section {
  name: string;
  start_measure: number;
  end_measure: number;
  notes: Note[];
  repeat: number;  // Number of times to play this section (1 = no repeat, 2 = repeat once)
  playback_notes?: Note[];  // Expanded notes for playback (repeats unrolled)
}

export interface Tune {
  id: string;
  title: string;
  key: string;
  time_signature: string;
  default_tempo: number;
  abc: string;
  sections: Section[];
  pickup_beats: number;  // Duration of pickup notes in beats (0 if no pickup)
}

export interface TuneSummary {
  id: string;
  title: string;
}

export interface TuneInfo extends TuneSummary {
  source?: string;
  source_url?: string;
  quality?: string;
  rating?: number | null;
  rating_count: number;
  owner?: string;
  version?: number;
}

export type PlaybackState = 'stopped' | 'playing' | 'paused';

export type SectionMode = 'full' | 'A' | 'B';

export type TuneSource = 'library' | 'user_song';

export interface SetItem {
  id: number;
  position: number;
  tune_source: TuneSource;
  tune_ref: string;
  tune_title: string;
  bpm: number | null;
  transpose: number | null;
  octave_shift: number | null;
  synth_type: string | null;
  metronome_enabled: boolean | null;
  count_off_enabled: boolean | null;
}

export interface SetSummary {
  id: number;
  name: string;
  item_count: number;
  updated_at: string;
}

export interface SetDetail {
  id: number;
  name: string;
  items: SetItem[];
  created_at: string;
  updated_at: string;
}

File: backend/tune.py
"""Pydantic models for tune data."""

from pydantic import BaseModel


class Note(BaseModel):
    """A single note with pitch, duration, and timing."""
    pitch: str  # e.g., "D4", "F#4"
    duration: float  # in beats
    start_time: float  # in beats from section start


class Section(BaseModel):
    """A section of a tune (e.g., A part, B part)."""
    name: str  # e.g., "A", "B"
    start_measure: int
    end_measure: int
    notes: list[Note]
    repeat: int = 1  # Number of times to play this section (1 = no repeat, 2 = repeat once)
    pickup_beats: float = 0.0  # Duration of pickup notes in this section (0 if no pickup)
    playback_notes: list[Note] | None = None  # Expanded notes for playback (repeats unrolled)


class Tune(BaseModel):
    """Complete tune data."""
    id: str
    title: str
    key: str
    time_signature: str
    default_tempo: int
    abc: str
    sections: list[Section]
    pickup_beats: float = 0.0  # Duration of pickup notes in beats (0 if no pickup)


class TuneSummary(BaseModel):
    """Summary info for tune list."""
    id: str
    title: str
    key: str
